GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs

```diff
@@ -0,0 +1,174 @@
+
+            return FixResult.Success;
+        }
+
+        /// <summary>
+        /// Get the list of full symbolic references (eg. HEAD, refs/remotes/origin/HEAD, refs/heads/master)
+        /// to inspect for and correct issues.
+        /// </summary>
+        /// <returns></returns>
+        protected abstract IEnumerable<string> GetRefs();
+
+        /// <summary>
+        /// Check if the contents of a reference is valid.
+        /// </summary>
+        /// <param name="fullSymbolicRef">A full symbolic ref name. eg. HEAD, refs/remotes/origin/HEAD, refs/heads/master</param>
+        /// <param name="refContents">Contents of the ref file</param>
+        protected virtual bool IsValidRefContents(string fullSymbolicRef, string refContents)
+        {
+            // Check for symbolic references
+            const string MinimallyValidHeadRef = "ref: refs/";
+            if (refContents.StartsWith(MinimallyValidHeadRef, StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+
+            // Otherwise check for raw commit-style references
+            return SHA1Util.IsValidShaFormat(refContents);
+        }
+
+        /// <summary>
+        /// 'git ref-log' doesn't work if the repo is corrupted, so parsing reflogs seems like the only solution.
+        /// </summary>
+        /// <param name="fullSymbolicRef">A full symbolic ref name. eg. HEAD, refs/remotes/origin/HEAD, refs/heads/master</param>
+        private static bool TryReadLastRefLogEntry(Enlistment enlistment, string fullSymbolicRef, out RefLogEntry refLog, out string error)
+        {
+            string refLogPath = Path.Combine(enlistment.WorkingDirectoryRoot, GVFSConstants.DotGit.Logs.Root, fullSymbolicRef);
+            if (!File.Exists(refLogPath))
+            {
+                refLog = null;
+                error = $"Could not find reflog for ref '{fullSymbolicRef}'";
+                return false;
+            }
+
+            try
+            {
+                string refLogContents = File.ReadLines(refLogPath).Last();
+                if (!RefLogEntry.TryParse(refLogContents, out refLog))
+                {
+                    error = $"Last ref log entry for '{fullSymbolicRef}' is unparsable.";
+                    return false;
+                }
+            }
+            catch (IOException ex)
+            {
+                refLog = null;
+                error = $"IOException while reading reflog '{fullSymbolicRef}': " + ex.Message;
+                return false;
+            }
+
+            error = null;
+            return true;
+        }
+
+        private bool TryParseRef(string fullSymbolicRef, List<string> messages = null)
+        {
+            string refPath = Path.Combine(this.Enlistment.WorkingDirectoryRoot, GVFSConstants.DotGit.Root, fullSymbolicRef);
+            if (!File.Exists(refPath))
+            {
+                messages?.Add($"Could not find ref file for '{fullSymbolicRef}'");

```

Comment:
I would indicate what path you were looking for here.

I just tried testing this code and my first attempt hit a failure to repair. My repro was to simply open up `.git/refs/heads/master` and delete its contents. The output tells me that it can't find the ref, but I'm not sure why not because it's there.

```
.git\refs\heads\**\*
    Could not find ref file for 'master'
    Repair failed. Run 'gvfs log C:\GVFSRepos\GVFS' for more info.
    Could not find reflog for ref 'master'
    Not all references could be fixed. Failed to fix 1 references.
```

Reply:
I believe it's looking for a file named `.git/master`, rather than `.git/refs/heads/master`

Reply:
Ah yes. I've got my base path/ref name prefixes mixed up here. Will address in next push.